### PR TITLE
8272579: G1: remove unnecesary null check for G1ParScanThreadStateSet::_states slots

### PR DESCRIPTION
**Pull request by albertnetymk**
Simple change of replacing a null-check with an assertion.

Test: hotspot_gc

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8272579](https://bugs.openjdk.java.net/browse/JDK-8272579): G1: remove unnecesary null check for G1ParScanThreadStateSet::_states slots

### Reviewers
 * [Ivan Walulya](https://openjdk.java.net/census#iwalulya) (@walulyai - Committer) ⚠️ Review applies to 301f00870e1535f5c33c3c5364c586ec2feed8a6
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Stefan Johansson](https://openjdk.java.net/census#sjohanss) (@kstefanj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5145/head:pull/5145` \
`$ git checkout pull/5145`

Update a local copy of the PR: \
`$ git checkout pull/5145` \
`$ git pull https://git.openjdk.java.net/jdk pull/5145/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5145`

View PR using the GUI difftool: \
`$ git pr show -t 5145`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5145.diff">https://git.openjdk.java.net/jdk/pull/5145.diff</a>

</details>

### CODE REVIEW

**Review by walulyai — APPROVED** [review]

**Review by kimbarrett — APPROVED** [review]
Looks good.

File: src/hotspot/share/gc/g1/g1ParScanThreadState.cpp [hunk 1]
```diff
@@ -558,10 +558,7 @@ void G1ParScanThreadStateSet::flush() {
 
   for (uint worker_id = 0; worker_id < _n_workers; ++worker_id) {
     G1ParScanThreadState* pss = _states[worker_id];
-
-    if (pss == NULL) {
-      continue;
-    }
+    assert(pss != nullptr, "must be initialized");
```
Comment (kstefanj) [inline src/hotspot/share/gc/g1/g1ParScanThreadState.cpp:561]: Is it guaranteed that all workers are initialized? As you state in the bug, they are lazily initialized, so I wonder if there are use cases where a worker would not get initialized during the collection. Something like, very little work and a lot of workers. 

If there is a guarantee here, does it hold below in `record_unused_optional_region()` as well? We are using the same construct with `continue`there.

Comment (albertnetymk) [reply src/hotspot/share/gc/g1/g1ParScanThreadState.cpp:561]: In `G1CollectedHeap::evacuate_initial_collection_set`:

```cpp
G1ParScanThreadStateSet per_thread_states(this,
                                          &rdcqs,
                                          workers()->active_workers(), <---- setting up states for workers
                                          collection_set()->young_region_length(),
                                          collection_set()->optional_region_length());
...
evacuate_initial_collection_set(&per_thread_states, may_do_optional_evacuation);
\
 -- task_time = run_task_timed(&g1_par_task); <--- launching all workers
```
Therefore, I think the assertion is safe.

> does it hold below in record_unused_optional_region() as well?

I think so.

Comment (albertnetymk) [reply src/hotspot/share/gc/g1/g1ParScanThreadState.cpp:561]: Changed; and tier1-3 pass.

**Comment by albertnetymk** [comment]
I updated the JBS ticket description/title as well to reflect the new patch.

**Review by kimbarrett — APPROVED** [review]
Still looks good.

**Review by kstefanj — APPROVED** [review]
Looks good now.

**Comment by albertnetymk** [comment]
Thanks for the review.

/integrate

**Pull request closed without merging**